esp32_s2_kaluga_kit/esp32_s2_kaluga_kit.c

```diff
@@ -119,24 +156,84 @@ esp_err_t bsp_audio_init(const i2s_std_config_t *i2s_config, i2s_chan_handle_t *
         BSP_ERROR_CHECK_RETURN_ERR(i2s_channel_enable(*rx_channel));
     }
 
-    /* Setup power amplifier enable pin */
-    const gpio_config_t io_conf = {
-        .intr_type = GPIO_INTR_DISABLE,
-        .mode = GPIO_MODE_OUTPUT,
-        .pin_bit_mask = BIT64(BSP_POWER_AMP_IO),
-        .pull_down_en = GPIO_PULLDOWN_DISABLE,
-        .pull_up_en = GPIO_PULLDOWN_DISABLE,
+    audio_codec_i2s_cfg_t i2s_cfg = {
+        .port = I2S_NUM_0,
+        .rx_handle = *rx_channel,
+        .tx_handle = *tx_channel,
     };
-    BSP_ERROR_CHECK_RETURN_ERR(gpio_config(&io_conf));
+    i2s_data_if = audio_codec_new_i2s_data(&i2s_cfg);
+    BSP_NULL_CHECK(i2s_data_if, ESP_FAIL);
 
     return ESP_OK;
 }
 
-esp_err_t bsp_audio_poweramp_enable(bool enable)
+/**
+ * @brief Common codec init
+ *
+ * Kaluga kit uses one codec for both audio playback and recording
+ *
+ * @return esp_codec_dev_t
+ */
+static esp_codec_dev_handle_t bsp_audio_codec_init(void)
 {
-    BSP_ERROR_CHECK_RETURN_ERR(gpio_set_level(BSP_POWER_AMP_IO, enable ? 1 : 0));
+    static esp_codec_dev_handle_t codec = NULL;
+    if (codec == NULL) {
+        if (i2s_tx_chan == NULL || i2s_rx_chan == NULL || i2s_data_if == NULL) {
+            /* Initialize I2C and I2S */
+            BSP_ERROR_CHECK_RETURN_ERR(bsp_i2c_init());
+            BSP_ERROR_CHECK_RETURN_ERR(bsp_audio_init(NULL, &i2s_tx_chan, &i2s_rx_chan));
+        }
+        assert(i2s_data_if);
+
+        const audio_codec_gpio_if_t *gpio_if = audio_codec_new_gpio();
+
+        audio_codec_i2c_cfg_t i2c_cfg = {
+            .port = BSP_I2C_NUM,
+            .addr = ES8311_CODEC_DEFAULT_ADDR,
+        };
+        const audio_codec_ctrl_if_t *i2c_ctrl_if = audio_codec_new_i2c_ctrl(&i2c_cfg);
+        BSP_NULL_CHECK(i2c_ctrl_if, NULL);
+
+        esp_codec_dev_hw_gain_t gain = {
+            .pa_voltage = 5.0,
+            .codec_dac_voltage = 3.3,
+        };
+
+        es8311_codec_cfg_t es8311_cfg = {
+            .ctrl_if = i2c_ctrl_if,
+            .gpio_if = gpio_if,
+            .codec_mode = ESP_CODEC_DEV_WORK_MODE_BOTH,
+            .pa_pin = BSP_POWER_AMP_IO,
+            .pa_reverted = false,
+            .master_mode = false,
+            .use_mclk = true,
+            .digital_mic = false,
+            .invert_mclk = false,
+            .invert_sclk = false,
+            .hw_gain = gain,
+        };
+        const audio_codec_if_t *es8311_dev = es8311_codec_new(&es8311_cfg);
+        BSP_NULL_CHECK(es8311_dev, NULL);
+
+        esp_codec_dev_cfg_t codec_dev_cfg = {
+            .dev_type = ESP_CODEC_DEV_TYPE_IN_OUT,
+            .codec_if = es8311_dev,
+            .data_if = i2s_data_if,
+        };
+        codec = esp_codec_dev_new(&codec_dev_cfg);
+        BSP_NULL_CHECK(codec, NULL);
```

Comment:
The idea here is that depending on Kconfig option `CONFIG_BSP_ERROR_CHECK` this function either returns NULL or aborts, in case of error.

We could simply return NULL everytime, but then users who expect every failure to abort() will be surprised...

Reply:
I understand, you are right.